src/site/antora/modules/ROOT/pages/manual/customconfig.adoc

```diff
@@ -14,372 +14,232 @@
-    updateLoggers(appender, config);
-}
-
-private void updateLoggers(final Appender appender, final Configuration config) {
-    final Level level = null;
-    final Filter filter = null;
-    for (final LoggerConfig loggerConfig : config.getLoggers().values()) {
-        loggerConfig.addAppender(appender, level, filter);
-    }
-    config.getRootLogger().addAppender(appender, level, filter);
-}
+include::example$manual/customconfig/Usage.java[tag=combineConfigurations,indent=0]
 ----
+<1> Loading a common, and an application-specific configuration from file
+<2> Casting them to `AbstractConfiguration`, the type required by `CompositeConfiguration`
+<3> Programmatically creating an uninitialized configuration.
+Note that no casting is needed.
+<4> Creating a `CompositeConfiguration` using all three configurations created.
+Note that passed configuration order matters!
+
+.How does `CompositeConfiguration` work?
+[%collapsible]
+====
+include::partial$manual/composite-configuration.adoc[tag=how]
+====
 
-You can achieve the same effect with an `OutputStream`, like a
-`PrintStream`:
+[#Hybrid]
+=== [[AddingToCurrent]] [[AppendingToWritersAndOutputStreams]] Modifying configuration components
+
+[WARNING]
+====
+*We strongly advise against programmatically modifying components of a configuration!*
+This section will explain what it is, and why you should avoid it.
+====
+
+It is unfortunately common that users modify components (appenders, filters, etc.) of a configuration programmatically as follows:
 
 [source,java]
 ----
-void addAppender(final OutputStream outputStream, final String outputStreamName) {
-    final LoggerContext context = LoggerContext.getContext(false);
-    final Configuration config = context.getConfiguration();
-    final PatternLayout layout = PatternLayout.createDefaultLayout(config);
-    final Appender appender = OutputStreamAppender.createAppender(layout, null, outputStream, outputStreamName, false, true);
-    appender.start();
-    config.addAppender(appender);
-    updateLoggers(appender, config);
-}
+LoggerContext context = LoggerContext.getContext(false);
+Configuration config = context.getConfiguration();
+PatternLayout layout = PatternLayout.createDefaultLayout(config);
+Appender appender = createCustomAppender();
+appender.start();
+config.addAppender(appender);
+updateLoggers(appender, config);
 ----
 
-The difference is the use of `OutputStreamAppender` instead of
-`WriterAppender`.
+This approach is prone several problems:
+
+* Your code relies on Log4j Core internals which don't have any backward compatibility guarantees.
+You not only risk breaking your build at a minor Log4j Core version upgrade, but also make the life of Log4j maintainers trying to evolve the project extremely difficult.
+* You move out from the safety zone, where Log4j Core takes care of components' life cycle (initialization, reconfiguration, etc.), and step into a minefield seriously undermining the reliability of your logging setup.
```

Comment:
I think we could be more specific than that:

* the code will not fail to compile at a minor release, because we take BC (too?) seriously. We might however add new required arguments to the builders, which will make the `build()` call fail. More probably we could deprecate methods.
* we can explicitly state that by "reliability" we mean that during the execution of this code, log events might be lost, unless the method calls are carefully ordered. 

Personally I find building components directly a better and type-safe alternative to `ConfigurationBuilder`.
The only problem is that building a configuration from scratch should **not** be performed, while the configuration is active.
The code should only run in a configuration factory or the `doConfigure()` method of a custom configuration.
`Configurator.reconfigure()` can be used to swap the old and new configuration.

Reply:
I'll make a PR.

Reply:
I have seen #2702 and [commented on it](https://github.com/apache/logging-log4j2/pull/2702#pullrequestreview-2154646468). I don't see any rationale (why it is better, etc.), but just code. Sorry, I disagree with this proposal. Let's not encourage anyone on using `PatternLayout.createDefaultLayout()`.